src/darwin/Framework/CHIP/ServerEndpoint/MTRServerEndpoint.mm

```diff
@@ -0,0 +1,197 @@
+{
+    if (!(self = [super init])) {
+        return nil;
+    }
+
+    _endpointID = [endpointID copy];
+    _deviceTypes = [deviceTypes copy];
+    _accessGrants = [[NSMutableSet alloc] init];
+    _serverClusters = [[NSMutableArray alloc] init];
+    _matterAccessGrants = [NSSet set];
+
+    for (MTRAccessGrant * grant in accessGrants) {
+        // Since our state is MTRServerEndpointStateInitializing, we know this
+        // will succeed.
+        [self addAccessGrant:[grant copy]];
+    }
+
+    for (MTRServerCluster * cluster in clusters) {
+        // Since our state is MTRServerEndpointStateInitializing and the initial
+        // cluster array was valid, we know this will succeed.
+        [self addServerCluster:[cluster copy]];
+    }
+
+    return self;
+}
+
+- (void)updateMatterAccessGrants
+{
+    MTRDeviceController * deviceController = _deviceController;
+    if (deviceController == nil) {
+        // _matterAccessGrants will be updated when we get bound to a controller.
+        return;
+    }
+
+    NSSet * grants = [_accessGrants copy];
+    [deviceController asyncDispatchToMatterQueue:^{
+        self->_matterAccessGrants = grants;
+    }
+                                    errorHandler:nil];
+}
+
+- (void)addAccessGrant:(MTRAccessGrant *)accessGrant
+{
+    [_accessGrants addObject:accessGrant];
+
+    [self updateMatterAccessGrants];
+}
+
+- (void)removeAccessGrant:(MTRAccessGrant *)accessGrant;
+{
+    [_accessGrants removeObject:accessGrant];
+
+    [self updateMatterAccessGrants];
+}
+
+- (BOOL)addServerCluster:(MTRServerCluster *)serverCluster
+{
+    MTRDeviceController * deviceController = _deviceController;
+    if (deviceController != nil) {
+        MTR_LOG_ERROR("Cannot add cluster on endpoint %llu which is already in use", _endpointID.unsignedLongLongValue);
+        return NO;
+    }
+
+    if (serverCluster.parentEndpoint != kInvalidEndpointId) {
+        MTR_LOG_ERROR("Cannot add cluster to endpoint %llu; already added to endpoint %" PRIu32, _endpointID.unsignedLongLongValue, serverCluster.parentEndpoint);
+        return NO;
+    }
+
+    for (MTRServerCluster * existingCluster in _serverClusters) {
+        if ([existingCluster.clusterID isEqual:serverCluster.clusterID]) {
+            MTR_LOG_ERROR("Cannot add second cluster with ID %llx on endpoint %llu", serverCluster.clusterID.unsignedLongLongValue, _endpointID.unsignedLongLongValue);
+            return NO;
+        }
+    }
+
+    [_serverClusters addObject:serverCluster];
+    serverCluster.parentEndpoint = static_cast<EndpointId>(_endpointID.unsignedLongLongValue);
+    return YES;
+}
+
+- (BOOL)associateWithController:(MTRDeviceController *)controller
+{
+    MTRDeviceController * existingController = _deviceController;
+    if (existingController != nil) {
+#if MTR_PER_CONTROLLER_STORAGE_ENABLED
+        MTR_LOG_ERROR("Cannot associate MTRServerEndpoint with controller %@; already associated with controller %@",
```

Comment:
Can there be multiple objects representing the same MTRDeviceController now?

Reply:
There can't.  I'm not sure I understand the question here...

All I am doing is that if we have uniqueIdentifier I am logging it, to make it easier to see in the logs which controllers are involved.

Reply:
Discussed on Slack, and leaving this as-is.